File: services/tasks.js
const { google } = require('googleapis');

/**
 * Fetches pending (incomplete) tasks from the user's default Google Tasks list.
 *
 * @param {string} accessToken
 * @param {number} maxResults
 * @returns {Array} clean task objects
 */
async function getTasks(accessToken, maxResults = 20) {
  const auth = new google.auth.OAuth2();
  auth.setCredentials({ access_token: accessToken });

  const tasks = google.tasks({ version: 'v1', auth });

  const res = await tasks.tasks.list({
    tasklist:      '@default',
    maxResults,
    showCompleted: false,
  });

  return (res.data.items || []).map(t => ({
    id:     t.id,
    title:  t.title,
    notes:  t.notes || '',
    due:    t.due || '',
    status: t.status,
  }));
}

/**
 * Creates a new task in the user's default Google Tasks list.
 * This is the "agentic execution" piece — Mimo writing back to Google, not just reading.
 *
 * @param {string} accessToken
 * @param {string} title
 * @param {string} notes
 * @returns {object} the created task (raw Google Tasks API response)
 */
async function createTask(accessToken, title, notes = '') {
  const auth = new google.auth.OAuth2();
  auth.setCredentials({ access_token: accessToken });

  const tasks = google.tasks({ version: 'v1', auth });

  const res = await tasks.tasks.insert({
    tasklist:    '@default',
    requestBody: { title, notes },
  });

  return res.data;
}

/**
 * Formats tasks into a clean text block for Claude.
 */
function formatTasksForContext(taskList) {
  if (!taskList.length) return 'No pending tasks found.';

  return taskList.map((t, i) =>
    `Task ${i + 1}:
  Title: ${t.title}
  Due:   ${t.due || 'No due date'}
  Notes: ${t.notes || 'N/A'}`
  ).join('\n\n');
}

module.exports = { getTasks, createTask, formatTasksForContext };

File: services/tasks.py
import asyncio
from typing import List, Dict, Any, Optional
from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build

async def get_tasks(access_token: str, max_results: int = 20) -> List[Dict[str, Any]]:
    """
    Fetches pending (incomplete) tasks from the user's default Google Tasks list.
    """
    creds = Credentials(token=access_token)

    def fetch_sync() -> List[Dict[str, Any]]:
        service = build("tasks", "v1", credentials=creds, cache_discovery=False)
        
        res = service.tasks().list(
            tasklist="@default",
            maxResults=max_results,
            showCompleted=False
        ).execute()
        
        tasks_list = res.get("items", [])
        
        return [{
            "id": t.get("id"),
            "title": t.get("title"),
            "notes": t.get("notes", ""),
            "due": t.get("due", ""),
            "status": t.get("status")
        } for t in tasks_list]

    return await asyncio.to_thread(fetch_sync)


async def create_task(access_token: str, title: str, notes: str = "") -> Dict[str, Any]:
    """
    Creates a new task in the user's default Google Tasks list.
    This is the "agentic execution" piece — Mimo writing back to Google, not just reading.
    """
    creds = Credentials(token=access_token)

    def insert_sync() -> Dict[str, Any]:
        service = build("tasks", "v1", credentials=creds, cache_discovery=False)
        
        task_body = {
            "title": title,
            "notes": notes
        }
        
        res = service.tasks().insert(
            tasklist="@default",
            body=task_body
        ).execute()
        
        return res

    return await asyncio.to_thread(insert_sync)


def format_tasks_for_context(task_list: List[Dict[str, Any]]) -> str:
    """
    Formats tasks into a clean text block for the LLM.
    """
    if not task_list:
        return "No pending tasks found."

    formatted_list = []
    for i, t in enumerate(task_list):
        task_block = (
            f"Task {i + 1}:\n"
            f"  Title: {t.get('title')}\n"
            f"  Due:   {t.get('due') or 'No due date'}\n"
            f"  Notes: {t.get('notes') or 'N/A'}"
        )
        formatted_list.append(task_block)

    return "\n\n".join(formatted_list)